Added functionnalities to mid node (fabric-client 1.4.0)

Andromelus opened the pull request:
# fabric-config node
Identity name and display name of the configuration are separated, allowing the user to insert a custom configuration name. This is useful in case the user uses identities from two organizations but with the same name (org1 admin and org2 admin for example). Previously, we could only see two "admin" configurations, forcing us to open the configuration to check which config correspond to which org.

# fabric-mid node

## Dynamic configuration
This node now handles dynamic configuration. It can be configured via the `msg.payload` object. 
Any configuration parameter passed to the msg.payload object will overwrite any manual configuration of the node.
For the full list of the configuration parameters, depending on what action the node must perform, please check node description

## Submit and Evaluate transaction
No change. They still use `fabric-network 1.4.0`

## Listen for event(s)

It uses `fabric-client 1.4.0` 

This node can listen for events. There are two "modes" of listening. This mode can be set with the timeout configuration parameter.

The default mode is `timeout = true;`

### Listen for ever

`timeout = false`

If timeout is false, the listener never disconnect/unregister and keep listening. All the events are sent one by one to the next node in `msg.payload`.

**Warning** : If you specified an interval with an end block, the listener **WILL NOT** disconect or unregister. If an interval with an end block is specified, it is preferable to set `timeout = true`.

### Listen "not for ever"

This is the default mode, with `timeout = true`.

The listener reset every time a transaction is found. The timeout's value is 2 seconds.

The timeout is needed because of a bug with the listener and because of the way it works.
More information here: https://jira.hyperledger.org/browse/FABN-1207

The events are sent to the next node all together in an array in `msg.payload`.

## Query world state

The mid node can now query the world state. This uses `fabric-client 1.4.0`.

## Query block

The mid node can now query a block in the ledger. Block is queried by block number.
This uses `fabric-client 1.4.0`.

## Query transaction

The mid node can now query a transaction in the ledger. Transaction is queried by transaction ID.
This uses `fabric-client 1.4.0`.

Commit:
https://github.ibm.com/CSOL/blockchain-code/issues/64

Initial commit, setup.

Commit:
#https://github.ibm.com/CSOL/blockchain-code/issues/64

Created an event listener that returns the evnet data + block number + txid, this event listener can start from a specific block up to the newest one.
Created a function to connect to the peer

Commit:
https://github.ibm.com/CSOL/blockchain-code/issues/64

Can specify an end block.
However, there is still a TODO.

Commit:
https://github.ibm.com/CSOL/blockchain-code/issues/64 and https://github.ibm.com/CSOL/blockchain-code/issues/65

Listed places where we could modify the existing node-red node to get block number and customize start/end block

Commit:
https://github.ibm.com/CSOL/blockchain-code/issues/64

Added a function to perform queries (simple and rich)

Commit:
Nodes modified

Need to apply proper tests. Need to clean/comment

Commit:
Query node updated | Labels updated

Query node: When data are provided via msg.payload, overwrites manual node config (except for connection profile)
Labels: Nodes now show in/mid/out/query labels to distinguish them.

Commit:
WIP - Rewrite own fabric-in (manual config)

Commit:
WIP - Event list node

New node that has an input/output for event listening.
Should be compatible with preexisting nodes (not fully tested)
Uses different connection method
Todo: Implement dynamic configuration with msg.payload

Commit:
WIP - Cleaned code | Backed up in node at its original state for clarity

Todo: Put a startblock in the in node

Commit:
WIP - Custom event node

Implemented choice between timeout or not
Purpose: This node can be used as a temp listener or as a continuous listener. It can also take a start block, allowing to re-listen if server falls

Commit:
Fields re not required anymore as they can be set by input

Commit:
Cleaned some console.log

Commit:
Update README.md

Commit:
Removed test code

Commit:
Added a label to the config node

Commit:
Added dynamic event filtering (not tested enough, but seems functionnal sur far)

Commit:
Fix - Return full msg object. Only payload is overwritten, other properties remain intact

Commit:
Fix - return full msg with payload in timeout

Commit:
Changed version for tagging.

Purpose: Refacto code + Redefine needs (rushed it too much)

Commit:
Removed filter from event listener, added information in readme about the purpose of these modifications

Commit:
Removed useless variable

Commit:
Changed label name to avoid confusion

Commit:
Mid node config can be dynamic. manual config by default, msg.payload if given

Commit:
Upated readme, set timeout from 3 to 2 seconds

Commit:
Update fabric.js

Commit:
Merge pull request #1 from floriancastelain/develop

Develop

Commit:
V0.0.3, removed query node to merge the functionalities with the mid node

Commit:
Merges nodes to keep only three with 1.4 and keeps fabric-client (fabric-network only in 1.4.1)

1.4.1 has its own branch that is developed on side because of lack of information about the sdk.

Commit:
Added transaction and block query functions

WARN: Need to clean code and write doc

Commit:
Added the possibility to query block by number and transaction by txid

Commit:
variable names, variable needs changed

actionTypeForm => actionType
connectToPeer => Don't need to pass a wallet location anymore, it is built from connection wallet path and identity name

Commit:
Simplified argument for event listener (fabric-client)

Commit:
Update README.md

Andromelus-ng committed:
Merge pull request #2 from floriancastelain/dev-1.4.0

Dev 1.4.0

Commit:
Fix eslint

Andromelus-ng committed:
Merge pull request #3 from floriancastelain/dev-1.4.0

Fix eslint

Commit:
Fixed variable name mistake

Commit:
another variable mistake...

Andromelus-ng committed:
Merge pull request #4 from floriancastelain/dev-1.4.0

Dev 1.4.0

cazfletch commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
     "name": "node-red-contrib-fabric",
-    "version": "0.0.1",
+    "version": "0.0.3",
```
why 0.0.3? shouldn't this be 0.0.2

Andromelus replied:
Mistake from me.

cazfletch commented on nodes/fabric.js:
```diff
@@ -74,12 +74,35 @@ module.exports = function (RED) {
+     * @param {Node} node node
+     * @returns {Promise<Buffer>} promise
+     */
+    function query(channel, payload, contractName, node) {
```
you don't need this it does the same as evaluate

Andromelus replied:
True. I did not pay attention to that. I'll remove the query.

cazfletch commented on nodes/fabric.js:
```diff
@@ -121,32 +147,166 @@ module.exports = function (RED) {
+     * @param {object} msg the msg object
+     * @returns {Promise<Buffer>} promise
+     */
+    async function subscribeToEvent(peer, channel, chaincodeName, node, msg) {
```
I think you can combine the two functions. They both do the same thing

Andromelus replied:
That is what I wanted to do first, but they work in a slightly different context. The original one (yours) starts right when the flow starts (in node). However, the version I made starts only when it receives a message from the precedent node. I've not checked that in details yet. I'll come back to you if I have issues with it.

cazfletch commented on nodes/fabric.js:
```diff
@@ -121,32 +147,166 @@ module.exports = function (RED) {
+     * @param {string} walletLocation the wallet location (the files, not the folder)
+     * @returns {PromiseLike<Contract | never>} promise
+     */
+    async function connectToPeer(identityName, channelName,
```
you don't need this function. You can get the channel from the network. You don't need the peer as you don't need to create a new event hub you can just get them like the subscribeToEvents function does, and you already have the identity from the wallet

Andromelus replied:
Correct. I'll remove it and use yours.

cazfletch reviewed (changes requested):
Thanks for the PR. You have some duplicate functionality and also are making it more complicated than it needs to be

Commit:
FIXED METADATA AND REMOVED QUERY

Set the version to 0.0.2 in package.json
Removed the query elements from fabric code and from UI

Commit:
EVENT HUBS FACTORY

Built an event hub factory that creates a channel event hub for each event listener.

Commit:
EVENT MANAGEMENT EVOLVED

- CC Event Listener: Built from a factory function
- Event Hub handler: Expected timed out hubs handled
- Handle two listener confs:
  - default: listen on all peers
  - peerName provided: listen only on this peer
- Updated nodes description

Commit:
ESLINT PASSED

Upgraded ECMA version to 8 (2017) because the code contains async.
Fixed code to match ESLINT. Note that one var is left because I need a "bigger" scope than let/const.

Andromelus-ng committed:
Merge pull request #5 from floriancastelain/dev-1.4.0

Dev 1.4.0

cazfletch commented on nodes/fabric.html:
```diff
@@ -50,140 +48,262 @@
-        <input type="text" id="node-input-eventName" placeholder=".*"/>
+        <input type="text" id="node-input-eventName" placeholder=".*" />
+    </div>
+    <div class="form-row">
```
can you make it obvious that these properties are optional see https://nodered.org/docs/creating-nodes/help-style-guide. You could also maybe make them in a twisty?

cazfletch reviewed (changes requested):
Looks good apart from the optional settings on the input node

Commit:
IN DESCRIPTION

Updated fabric-in description.

Andromelus-ng committed:
Merge pull request #6 from floriancastelain/dev-1.4.0

IN DESCRIPTION

cazfletch commented on nodes/fabric.html:
```diff
@@ -238,23 +239,54 @@
-                <li> If the <code>peerName</code> value is set, the node will listen only on the peer with the given name.</li>
-            </ul>
-        </li>
+    <p>Hyperledger Fabric input node. Subscribe to events emitted by a blockchain</p>
```
I'm not sure its obvious enough that these are optional, there is nothing on the node so you have to read the info

Commit:
HIGHLIGHTED OPTIONAL PARAMS FABRIC-IN

Andromelus-ng committed:
Merge pull request #7 from floriancastelain/dev-1.4.0

HIGHLIGHTED OPTIONAL PARAMS FABRIC-IN
